ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+    a slight deviation from other SDKs.
+* `Temporalio::Converter::PayloadConverter::Encoding` - Base unimplemented encoding converter
+  * ❓ Would we prefer `Temporalio::Converter::EncodingPayloadConverter`?
+* `Temporalio::Converter::PayloadConverter::JSON`, `Temporalio::Converter::PayloadConverter::JSONProto`, etc  - All
+  common encoding converters
+* `Temporalio::Converter::FailureConverter` - Default failure converter, can be overridden
+* `Temporalio::Converter::PayloadCodec` - Unimplemented base codec
+  * Important for implementers to ensure this is async capable if using async activities anywhere
+* `Temporalio::Converter::RawValue` - Can be used as way to move around unconverted payloads
+
+#### Temporalio::Converter::PayloadConverter::JSON
+
+* Ruby has a `JSON` module in the standard library, but by default it's only primitives, arrays, hashes, etc and not
+  classes.
+  * There is a concept of "additions", but they set a `json_class` field on the object which makes the JSON not
+    cross-language compatible.
+  * There are not any good JSON-to-object conversion approaches in Ruby beyond this.
+  * Therefore, very unfortunately, it is unreasonable at this time for any area where conversion from payload to value
+    is needed to specify the type to convert to. Users can use type hints in their RBS files and such, but that has no
+    runtime check/effect.
+* The JSON encoding converter initializer will have a way to set parse and generate options.
+* The payload converter will have a way to let the default be created with JSON parse and generate options for the JSON
+  encoding converter.
+  * 💭 Why? Users wanting to simply set JSON options shouldn't have to go through the mess of recreating the default
+    list of encoding converters.
+
+## Activities
+
+A simple, heartbeating activity that calls a separate thing:
+
+```ruby
+class MyActivity < Temporalio::Activity
+  # Can customize name, e.g.
+  # activity_name 'custom-name'
+
+  # Can customize executor, e.g.
+  # activity_executor :fiber
+
+  # Can make an activity dynamic, e.g.
+  # activity_dynamic true
+
+  def initialize(db_client)
+    @db_client = db_client
+  end
+
+  def execute(arg)
+    loop do
+      db_result = @db_client.do_thing(arg)
+      if db_result
+        return db_result
+      end
+      sleep 1
+      Temporalio::Activity::Context.heartbeat
+    end
+  end
+end
+```
+
+* Activities are defined as classes.
+  * 💭 Why classes instead of methods/functions like every other SDK? There are tradeoffs to both approaches. As
+    classes, they are more easily referenced, a bit more easily typed in RBS, and metadata can be applied more easily at
+    the class level. This comes at a cost of not being able to easily share state across classes, but users can provide
+    something in the constructor multiple activities use. Open to considering activity methods instead.
```

Comment:
This wouldn't be for an MVP, but could we eventually provide both if we wanted to? It doesn't seem like there's an obvious trade-off winner here.

Another plus in favor of functions is they are lighter-weight to define; you don't have to make a class just to wrap what is effectively a function anyway.

Reply:
We can technically do this, and could technically provide activities as classes/interfaces/structs in every other SDK but we have tried to stick with a single way that is best with the language. I think we should only do one here.

Reply:
Yeah if we want to pick a single way, then I'm in favor of doing the most-flexible thing. Long as we have room to add the other way later if we learn that users are really opinionated about it.